functions/index.js

```diff
@@ -1,8 +1,361 @@
 const functions = require('firebase-functions');
+const fs = require('fs');
+const path = require('path');
+var https = require("https");
+var url = require('url');
 
-// Create and Deploy Your First Cloud Functions
-// https://firebase.google.com/docs/functions/write-firebase-functions
+// travelData is the sample data that we use to demo filtering on the frontend.
+const src = path.join(path.dirname(__filename), 'travel-data.json');
+const travelData = JSON.parse(fs.readFileSync(src, 'utf8'));
 
-exports.helloWorld = functions.https.onRequest((request, response) => {
- response.send("Hello from Firebase!");
+/**
+ * cors is HTTP middleware that sets an AMP-specific cross origin resource
+ * sharing HTTP header on our response.
+ *
+ * @param {Object} req - The HTTP request object.
+ * @param {Object} res - The HTTP resonse object.
+ * @param {Function} next - The next HTTP handler to execute.
+ */
+function cors(req, res, next) {
+  const host = process.env.NODE_ENV === 'production'
+    ? `https://${req.hostname}`
+    : `http://${req.hostname}:5000`
+
+  res.header('amp-access-control-allow-source-origin', host);
+
+  next();
+}
+
+/**
+ * Cast a value to an array.
+ *
+ * The value is coerced to an array, in the case that the provided value is
+ * null, undefined or an empty string, an empty array is returned.
+ * @param  {any} val - any value.
+ * @returns {array} an array.
+ */
+function castArray(val) {
+  if (Array.isArray(val)) {
+    return val;
+  }
+
+  if (val === undefined || val === '' || val === null) {
+    return [];
+  }
+
+  return [val];
+}
+
+/**
+ * filter returns the travel data that matches the given query.
+ *
+ * @param {Array} data - Array of objects containing the travel data to
+ * filter.
+ * @param {Object} query - An HTTP request query object.
+ * @return {Array} - An array of travelData objects.
+ */
+function filter(data, query) {
+  var results = [];
+  var push = true;
+
+  // omit every result that doesn't pass _every_ filter
```

Comment:
the easiest way to do this is with a switch case and a loop inside making this efficient and you can avoid all (most of) the booleans and conditionals inside the loop.

Reply:
I'm not completely sure what you mean here. Could you be more specific, please?

Reply:
@camelburrito I can't visualize how a `switch` statement could improve this logic. Please let us know more about what you are thinking.

Reply:
If you pass in a filter by criteria it would be easier. But this logic is is not bad either, lets leave it as is. Its just less readable.